File: Linked list/86. Partition List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode temp_s=new ListNode(0);
        ListNode temp_h=new ListNode(0);
        ListNode small=temp_s;
        ListNode high=temp_h;
        
        while(head!=null)
        {
            if (head.val<x)
            {
                small.next=head;
                small=small.next;
            }
            else
            {
                high.next=head;
                high=high.next;
            }
            head=head.next;
        }
        small.next=temp_h.next;
        high.next=null;
        return temp_s.next;
    }
}

File: Linked list/86. Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode *partition(ListNode *head, int x)
    {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode *dummyLeft = new ListNode(-1);
        ListNode *dummyRight = new ListNode(-1);

        ListNode *left = dummyLeft;
        ListNode *right = dummyRight;

        ListNode *temp = head;

        while (temp != NULL)
        {
            if (temp->val < x)
            {
                left->next = temp;
                left = left->next;
            }
            else if (temp->val >= x)
            {
                right->next = temp;
                right = right->next;
            }
            temp = temp->next;
        }

        left->next = dummyRight->next;
        right->next = NULL;

        return dummyLeft->next;
    }
};